graphql_compiler/cost_estimation/int_value_conversion.py

```diff
@@ -0,0 +1,125 @@
+computation in the integer domain, and transfer the computation back into the original domain.
+
+In order to be able to reason about value intervals and successor/predecessor values, we
+make sure these mappings to integers are increasing bijective functions.
+
+This kind of mapping is easy to do for int, uuid and datetime types, but not possible for other
+types, like string. When the need for other types arises, the precise interface for range
+reasoning can be defined and implemented separately for each type.
+"""
+import datetime
+from typing import Any
+from uuid import UUID
+
+import pytz
+
+from ..schema.schema_info import QueryPlanningSchemaInfo
+from .helpers import is_date_field_type, is_datetime_field_type, is_int_field_type, is_uuid4_type
+
+
+# UUIDs are defined in RFC-4122 as a 128-bit identifier. This means that the minimum UUID value
+# (represented as a natural number) is 0, and the maximal value is 2^128-1.
+MIN_UUID_INT = 0
+MAX_UUID_INT = 2 ** 128 - 1
+
+
+DATETIME_EPOCH_UTC = datetime.datetime(1970, 1, 1, tzinfo=pytz.utc)
+
+
+def field_supports_range_reasoning(
+    schema_info: QueryPlanningSchemaInfo, vertex_class: str, property_field: str
+) -> bool:
+    """Return whether range reasoning is supported. See module docstring for definition."""
+    return (
+        is_uuid4_type(schema_info, vertex_class, property_field)
+        or is_int_field_type(schema_info, vertex_class, property_field)
+        or is_datetime_field_type(schema_info, vertex_class, property_field)
+        or is_date_field_type(schema_info, vertex_class, property_field)
+    )
+
+
+def convert_int_to_field_value(
+    schema_info: QueryPlanningSchemaInfo, vertex_class: str, property_field: str, int_value: int
+) -> Any:
+    """Return the given integer's corresponding property field value.
+
+    See module docstring for details. The int_value is expected to be in the range of
+    convert_field_value_to_int.
+
+    Args:
+        schema_info: QueryPlanningSchemaInfo
+        vertex_class: str, name of vertex class to which the property field belongs.
+        property_field: str, name of property field that the value refers to.
+        int_value: int, integer value which will be represented as a property field value.
+
+    Returns:
+        Any, the given integer's corresponding property field value.
+
+    Raises:
+        ValueError, if the given int_value is outside the range of valid values for the given
+        property field.
+    """
+    if is_int_field_type(schema_info, vertex_class, property_field):
+        return int_value
+    if is_datetime_field_type(schema_info, vertex_class, property_field):
+        return DATETIME_EPOCH_UTC + datetime.timedelta(microseconds=int_value)
+    if is_date_field_type(schema_info, vertex_class, property_field):
+        return datetime.date.fromordinal(int_value)
+    elif is_uuid4_type(schema_info, vertex_class, property_field):
```

Comment:
This is a bit strange. For consistency, I'd say either make line 70 and 72 also be `elif`, or add a blank line here and make this a regular `if` since all previous ones return.